Avoid excessive memory allocations that cause GC work

gchoqueux opened the pull request:
## Description
add or use optional target to methods to avoid memory allocations
If target is defined then use this to write the result instead of allocating an objet.

## Motivation and Context
Avoid excessive memory allocations that cause GC work.

gchoqueux committed:
feature(core): avoid memory allocation to function declaration

peppsac commented on src/Core/Math/Ellipsoid.js:
```diff
@@ -55,11 +56,10 @@ Ellipsoid.prototype.setSize = function setSize(size) {
 };
 
-
+const k = new THREE.Vector3();
```
Allocating `k` outside of `cartographicToCartesian` isn't a good idea, since you're returning it at each call.

I'd prefer that you add a `target` param and allocate a return vector if missing.

gchoqueux replied:
updated

gchoqueux commented:
PR updated

autra commented on src/Core/Geographic/Projection.js:
```diff
@@ -42,7 +42,7 @@ Projection.prototype.getCoordWMTS_WGS84 = function getCoordWMTS_WGS84(tileCoord,
         return this.WMTS_WGS84ToWMTS_PM(tileCoord, bbox);
     } else if (tileMatrixSet === 'WGS84G') {
-        return [tileCoord];
+        return [tileCoord.clone()];
```
Why do you need to clone here?

autra replied:
Because of https://github.com/iTowns/itowns/pull/615/files#diff-fe4e9c46a7cb968cb268ab300bafff1cR114 ? 

If it's the case, the clone must be done in the calling function, because it is the one which knows it needs a new instance.

gchoqueux replied:
This function shouldn't return an object's reference
The user isn't supposed to know that are result can be changed via another object

gchoqueux replied:
maybe rename in `computeCoordWMTS`

autra replied:
Ok we only currently use this in cases where we need a new object, so we can leave it like that for now, it makes sense.

autra commented on src/Core/Geographic/Projection.js:
```diff
@@ -123,20 +125,22 @@ Projection.prototype.WGS84toWMTS = function WGS84toWMTS(bbox) {
+    bbox.center(center);
     var col = Math.floor((MathExt.PI + center.longitude(UNIT.RADIAN)) / uX);
     var row = Math.floor(nY - (MathExt.PI_OV_TWO + center.latitude(UNIT.RADIAN)) / uY);
+    target = target || new Extent('WMTS:WGS84G', 0, 0, 0);
```
nits: I'd prefer this line at the beginning of the function (it's clearer to read, makes the intent of `target` easier to spot).

gchoqueux replied:
I updated the PR
I moved in default parameter

autra replied:
I don't think you can do that: if I'm not mistaken, default parameter values are instanced once and for all at parsing time, so you'll share one instance for everyone here :-)

gchoqueux replied:
I verified, a new instance is returned on each call for default parameter value

autra commented on src/Core/Geographic/Extent.js:
```diff
@@ -362,11 +362,17 @@ Extent.prototype.intersect = function intersect(other) {
+        this._row = values[1];
+        this._col = values[2];
+    } else {
+        Object.keys(CARDINAL).forEach((key) => {
```
Note, we have currently 4 ways of constructing an Extent. In the future, I guess we want to support the same ways in this method (so there are still 2 ways missing, with an object like `{west, east, south, north}` and with 2 `Coordinates` objects).

autra commented on src/Renderer/ThreeExtended/OBB.js:
```diff
@@ -22,7 +22,20 @@ OBB.prototype.clone = function clone() {
     this.updateMatrixWorld(true);
 
-    this.pointsWorld = this._cPointsWorld(this._points());
+    if (!this.pointsWorld) {
```
Any reason why this couldn't be done in the constructor once and for all instead? `update` can potentially be called often.

gchoqueux replied:
because for the tile, `update` function is called each time DEM is updated

autra replied:
Ok, but that wouldn't prevent `this.pointsWorld` initialisation and all the `new THREE.Vector3` to go into the constructor, right? I mean: `this._cPointsWorld` call would stay here.

gchoqueux replied:
yes, but these points are only used for the horizon culling I wanted to keep this separation to be able to create them only when necessary

autra replied:
Yes, but OBB.update is called in `TileMesh.updateMatrixWorld` which is called in subdivideNode and is also called in the constructor. So anyway, these vector3 will get initialized when you do `new OBB` whether or not you're doing horizontal culling (they also get init in planar views). So initializing them in the constructor would be clearer and more natural.

gchoqueux replied:
ok

autra reviewed (changes requested):
Already looking good thanks!

If you want to push this to the far end, there are still 2 instances of `new Extent` without targets in `Projection.js`:
- `Projection.prototype.WMTS_WGS84ToWMTS_PM `, 

I think this one is important because the call chain goes up to `TileMesh.getCoordsForLayer`, which is used by some providers' `tileInsideLimits` methods, often called in `layer.update`. I expect some win here because the `new Extent` is called inside a [loop](https://github.com/iTowns/itowns/blob/master/src/Core/Scheduler/Providers/WMTS_Provider.js#L149) of [loops](https://github.com/iTowns/itowns/blob/master/src/Core/Geographic/Projection.js#L108).

It would require passing an array of targets of the right size somewhere.

- `Projection.prototype.getCoordsWMTS `: seemed to be only called by `getAllCoordsWMTS`: we could delete both of them, for the sake of completion.

gchoqueux commented:
> `Projection.prototype.WMTS_WGS84ToWMTS_PM`

This function is called only if coords haven't been computed, there's no point in passing a target

> It would require passing an array of targets of the right size somewhere

this size is precisely calculated in `WMTS_WGS84ToWMTS_PM`

autra commented:
> > Projection.prototype.WMTS_WGS84ToWMTS_PM
> 
> This function is called only if coords haven't been computed, there's no point in passing a target

Indeed it is cached in tile.wmtsCoords anyway, so no need to touch that.

gchoqueux committed:
fix(gis): possibility to set tiled crs

gchoqueux commented:
@autra this PR can be merged

autra commented on src/Core/Math/Ellipsoid.js:
```diff
@@ -55,39 +53,30 @@ Ellipsoid.prototype.setSize = function setSize(size) {
  * @param {number} position.y
  * @param {number} position.z
- * @returns {EllipsoidCoordinate} an object describing the coordinates on the reference ellipsoid, angles are in degree
+ * @param {Coordinate} target coordinate to copy result
```
nits: you can indicate that the parameter is optional (see http://usejsdoc.org/tags-param.html#optional-parameters-and-default-values)

autra reviewed (approved):
LGTM, thanks!

gchoqueux committed:
feature(core): add or use optional target to methods to avoid memory allocations

Avoid excessive memory allocations that cause GC work.
If target is defined then use this to write the result instead of
allocating an objet.
